### PR TITLE
remove mutex from trie operations

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- The trie accesses are serialized using a mutex on all of the public functions. While this assures that parallel calls will not result in data races, it also introduces a bottleneck.
  
## Proposed changes
- Remove the mutex from the public trie functions, and synchronize the accesses at the trie node level.
- Remove `GetAllHashes` func from the trie. This was not used, and it could lead to oom if used for a large trie.
- Refactor `get`, `update` and `delete` from the trie so that they can be called in parallel

## Testing procedure
- Unit tests
- System tests when all the PRs are merged in the feat branch

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: trie/extensionNode.go [hunk 1]
```diff
@@ -268,48 +240,59 @@ func (en *extensionNode) getEncodedNode() ([]byte, error) {
...
+func (en *extensionNode) isCollapsed() bool {
+	return en.child == nil && len(en.EncodedChild) != 0
+}
+
+func (en *extensionNode) resolveIfCollapsed(db common.TrieStorageInteractor) (node, error) {
+	en.childMutex.Lock()
+	defer en.childMutex.Unlock()
+
+	isChildCollapsed := en.child == nil && len(en.EncodedChild) != 0
+	if !isChildCollapsed {
```
Comment (sasurobert) [inline trie/extensionNode.go:252]: why not use en.isCollaped? exactly the same functions is wrote on top of this function.

File: trie/patriciaMerkleTrie.go [hunk 2]
```diff
@@ -249,12 +255,14 @@ func (tr *patriciaMerkleTrie) deleteBatch(data []core.TrieData) error {
 		return nil
 	}
 
-	if tr.root == nil {
+	rootNode := tr.GetRootNode()
+	if rootNode == nil {
```
Comment (sasurobert) [inline trie/patriciaMerkleTrie.go:259]: check.IfNil ?

File: trie/patriciaMerkleTrie.go [hunk 3]
```diff
@@ -662,23 +668,22 @@ func logMapWithTrace(message string, paramName string, hashes common.ModifiedHas
...
-	err := tr.updateTrie()
-	if err != nil {
-		return nil, nil, err
-	}
+	tr.updateTrieMutex.Lock()
+	defer tr.updateTrieMutex.Unlock()
 
-	if tr.root == nil {
+	rootNode := tr.GetRootNode()
+	if rootNode == nil {
```
Comment (sasurobert) [inline trie/patriciaMerkleTrie.go:678]: check.ifnil?

File: trie/rootManager.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+type rootManager struct {
+	root         node
+	oldHashes    [][]byte
+	oldRootHash  []byte
+	mutOperation sync.RWMutex
+}
+
+func NewRootManager() *rootManager {
```
Comment (sasurobert) [inline trie/rootManager.go:15]: a lot of comments are missing here. on all the file.

File: trie/interface.go [hunk 5]
```diff
@@ -110,3 +107,12 @@ type IdleNodeProvider interface {
 	IsIdle() bool
 	IsInterfaceNil() bool
 }
+
+type RootManager interface {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/interface.go:111]: Added.

File: trie/extensionNode.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply trie/extensionNode.go:252]: Changed.

File: trie/patriciaMerkleTrie.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie.go:259]: Done.

File: trie/patriciaMerkleTrie.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie.go:678]: Done.

File: trie/rootManager.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply trie/rootManager.go:15]: Added

File: trie/branchNode.go [hunk 6]
```diff
@@ -683,43 +625,44 @@ func (bn *branchNode) delete(
...
+	return bn.reduceNodeIfNeeded(goRoutinesManager, modifiedHashes, db)
+}
+
+func (bn *branchNode) reduceNodeIfNeeded(
+	goRoutinesManager common.TrieGoroutinesManager,
+	modifiedHashes common.AtomicBytesSlice,
+	db common.TrieStorageInteractor,
+) (bool, node) {
+	bn.mutex.Lock()
+	defer bn.mutex.Unlock()
```
Comment (ssd04) [inline trie/branchNode.go:637]: if setters in baseNode are protected by mutex, i think this protection here is not needed

File: trie/interface.go [hunk 7]
```diff
@@ -15,32 +15,33 @@ import (
...
 type node interface {
 	getHash() []byte
-	setHash() error
 	setGivenHash([]byte)
+	setDirty(bool)
+	isDirty() bool
+	getMarshalizer() marshal.Marshalizer
+	setMarshalizer(marshal.Marshalizer)
+	getHasher() hashing.Hasher
+	setHasher(hashing.Hasher)
```
Comment (ssd04) [inline trie/interface.go:24]: maybe we can have a baseNode interface

File: trie/branchNode.go [hunk 8]
```diff
@@ -754,25 +697,31 @@ func (bn *branchNode) deleteChild(
...
+func (bn *branchNode) setNewChild(childPos byte, newNode node, modifiedHashes common.AtomicBytesSlice) error {
 	bn.mutex.Lock()
 	defer bn.mutex.Unlock()
 
+	if !bn.dirty && len(bn.hash) != 0 {
+		modifiedHashes.Append([][]byte{bn.hash})
+	}
+
 	bn.hash = nil
+	bn.dirty = true
```
Comment (ssd04) [inline trie/branchNode.go:709]: if mutex protection is added in baseNode, cand we use setters here and avoid mutex in this file entirely?

File: trie/leafNode.go [hunk 9]
```diff
@@ -201,23 +173,10 @@ func (ln *leafNode) getEncodedNode() ([]byte, error) {
...
-	return false
-}
-
 func (ln *leafNode) tryGet(key []byte, currentDepth uint32, _ common.TrieStorageInteractor) (value []byte, maxDepth uint32, err error) {
-	err = ln.isEmptyOrNil()
-	if err != nil {
-		return nil, currentDepth, fmt.Errorf("tryGet error %w", err)
-	}
+	ln.mutex.RLock()
+	defer ln.mutex.RUnlock()
```
Comment (ssd04) [inline trie/leafNode.go:178]: do we need mutex here?

Comment (ssd04) [reply trie/leafNode.go:178]: this mutex is protecting `hash` and `dirty` fields, right?

File: trie/modifiedChildren.go
Comment (ssd04) [inline trie/modifiedChildren.go:1]: add unit tests for this file (or TODO for future PRs)

File: trie/rootManager.go
Comment (ssd04) [inline trie/rootManager.go:1]: add unit tests for this file (or TODO for future PRs)

File: trie/baseNode.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+func (bn *baseNode) getHash() []byte {
+	return bn.hash
+}
+
+func (bn *baseNode) setGivenHash(hash []byte) {
+	bn.hash = hash
+}
+
+func (bn *baseNode) isDirty() bool {
+	return bn.dirty
```
Comment (ssd04) [inline trie/baseNode.go:26]: i think rootManager is accessing this so it needs protection here

**Review by ssd04 — CHANGES_REQUESTED** [review]

File: state/accountsDB.go [hunk 11]
```diff
@@ -534,44 +532,7 @@ func (adb *AccountsDB) removeCodeAndDataTrie(acnt vmcommon.AccountHandler) error
...
-	adb.obsoleteDataTrieHashes[string(rootHash)] = hashes
-
-	entry, err := NewJournalEntryDataTrieRemove(rootHash, adb.obsoleteDataTrieHashes)
-	if err != nil {
-		return err
-	}
-	adb.journalize(entry)
-
-	return nil
+	return adb.removeCode(baseAcc)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/accountsDB.go:535]: We need to keep track of this. When `RemoveAccount()` will be used, we need to implement an efficient way of marking the removed hashes for pruning. Added a TODO

File: trie/patriciaMerkleTrie.go [hunk 12]
```diff
@@ -201,22 +205,25 @@ func (tr *patriciaMerkleTrie) insertBatch(sortedDataForInsertion []core.TrieData
 		return nil
 	}
 
-	if tr.root == nil {
+	rootNode := tr.GetRootNode()
+	if rootNode == nil {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie.go:209]: Done.

File: trie/baseNode.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"github.com/multiversx/mx-chain-core-go/hashing"
+	"github.com/multiversx/mx-chain-core-go/marshal"
+	"sync"
+)
+
+type baseNode struct {
+	mutex  sync.RWMutex
+	hash   []byte
+	dirty  bool
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/baseNode.go:12]: Indded. Added TODO and will add mutex protection in a future PR (to minimize conflicts with already opened PRs)

File: trie/branchNode.go (hunk 6 above)
Comment (BeniaminDrasovean) [reply trie/branchNode.go:637]: It is needed in order to keep all changes atomic. If I remove this there might be edgecases where different baseNode fields change from different goroutines.

File: trie/interface.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply trie/interface.go:24]: Done.

File: trie/leafNode.go (hunk 9 above)
Comment (BeniaminDrasovean) [reply trie/leafNode.go:178]: It protects everything that can change on the node, so `ln.Key` & `ln.Value` also

File: trie/rootManager.go
Comment (BeniaminDrasovean) [reply trie/rootManager.go:1]: Added TODO

File: trie/baseNode.go (hunk 10 above)
Comment (BeniaminDrasovean) [reply trie/baseNode.go:26]: Indeed, will refactor mutex usage in a future PR

**Review by sasurobert — APPROVED** [review]

**Pull request merged**